Finaliza processo de importação de planilhas com validação de par + notificação

berinhard opened the pull request:
Fixes #212 
Fixes #239 

@turicas já setei as variáveis de ambiente no RocketChat no Dokku

turicas committed:
Atualiza PostgreSQL conf de prod

berinhard committed:
Reabre o arquivo XLS antes da importação

berinhard committed:
Move código de queries para Covid19Stats

berinhard committed:
Move custom exception para um arquivo de exceções da app

berinhard committed:
Implementa lógica de alto nível na tarefa assíncrona para viabilizar API e métodos a serem implementados

berinhard committed:
Evita problemas com difença de caixa alta/baixa na sigla dos estados pro caminho do arquivo da importação

berinhard committed:
Cria metódo que compara duas planilhas

berinhard committed:
Implementa lógica que compara uma única planilha aos seus pares

berinhard committed:
Recupera dados de planilha de links por UF

berinhard committed:
Adiciona dependência com o RocketChat à aplicação

berinhard committed:
Função auxiliar para recuperar o canal do rocket chat

berinhard committed:
Usa implementação customizada do RocketChat

berinhard committed:
Adiciona notificação de nova planilha importada

berinhard committed:
Implementa notificação de dados divergentes da planilha

berinhard committed:
Garantia dupla de que não estamos checando com planilha importada pelo próprio user

berinhard committed:
Durante a validação da planilha, ignora as linhas de cidade que tenham valores nulos para ambos os campos

berinhard committed:
Se os dados históricos são zerados/nulos e não presentes na importação, são dinamicamente adicionados

berinhard committed:
Marca a planilha como DEPLOYED no final do processo

turicas committed:
Separa http_get_json em 2 funções

turicas committed:
Gera planilha no modelo do Google Spreadsheets

turicas committed:
Merge branch 'develop' into feature/import-notification

turicas commented on covid19/notifications.py:
```diff
@@ -0,0 +1,69 @@
+
+    errors = "\n- ".join(errors)
+    msg = f"{collabs} - *Dados divergentes* na planilha importada para o estado *{spreadsheet.state}* para o dia *{spreadsheet.date}*"
+    msg += f"\nRealizada por: *{spreadsheet.user.username}*\n\n- {errors}"
```
Aqui é importante deixar claro que os dados só poderão ser atualizados se alguém submeter uma nova planilha e os dados baterem.

berinhard replied:
Resolvido em 11d9642

turicas commented on covid19/spreadsheet_validator.py:
```diff
@@ -75,11 +56,11 @@ def format_spreadsheet_rows_as_dict(rows_table, date, state):
         deaths = getattr(entry, deaths_attr, None)
 
-        if not confirmed and deaths or not deaths and confirmed:
+        if (not confirmed and deaths) or (not deaths and confirmed):
```
Verificar aqui sobre o `None` vs `0` (usar `v is None`/`v is not None` em vez de `not v`).

berinhard replied:
Resolvido em b2bf207

turicas commented on covid19/models.py:
```diff
@@ -135,3 +146,50 @@ def get_total_data(self):
+        return False, errors
+
+    def import_to_final_dataset(self):
+        self.status = StateSpreadsheet.DEPLOYED
```
Aqui tem que gerar uma notificação em `#covid19` me marcando e falando que esse estado já está checado e pronto para atualização.

berinhard replied:
Resolvido em ac0c02921fac915d730b0c71115601cb875ab813

turicas commented on covid19/models.py:
```diff
@@ -135,3 +146,50 @@ def get_total_data(self):
+        if errors:
+            return errors
+
+        for entry in self.table_data:
```
Se `other_spreadsheet` (sugiro usar só `other`) tiver dados de municípios diferentes que `self`, esse check não vai pegar.
Sugiro fazer a comparação assim:
- Gera uma property em `StateSpreadsheet` que te devolva um dicionário onde as chaves são o `place_key` e os valores sejam os dados para aquele local, assim esse dicionário contemplará tanto os registros de municípios quanto o total no estado (vou chamar aqui de `self.X` e `other.X`);
- Compara `sorted(self.X.keys())` com `sorted(other.X.keys())` - se forem diferentes, gera N erros:
  - `for err1 in set(self.X.keys()) - set(other.X.keys()): errors.append(f"{err1} está aqui e não lá")`
  - `for err2 in set(other.X.keys()) - set(self.X.keys()): errors.append(f"{err2} está lá e não aqui")`
- Para cada key em `self.X.keys()`, verifica se o valor dessa key é igual a `other.X[key]`;

berinhard replied:
Resolvido em be5d8d7

berinhard committed:
Melhora mensagem para colaboradoras

berinhard committed:
Checa se valores são nulos ao invés de falsos

berinhard committed:
Trata exceção pro caso de conjunto de cidades diferentes nas planilhas

berinhard committed:
Cria campo para poder recuperar o par de planilhas usadas na validação

berinhard committed:
Garante que a ligação entre o par de planilhas validadas é feito

berinhard committed:
Renomeia método para ficar mais coerente com seu funcionamento interno

berinhard committed:
Garante que só importamos planilhas que estejam de fato prontas para a importação

berinhard committed:
Valida planilha considerando somente planilhas que não possuem erros

berinhard committed:
Notifica @turicas no canal #covid19 após validação

Exemplo de validação que fiz no terminal:

In [5]: notify_import_success(spreadsheet)
New message in #covid19:
	@turicas planilha de PR para o dia 2020-04-21 checada (dados enviados por admin e bernardo), pode rodar o deploy!

berinhard committed:
Deve retornar uma instância de FakeChat e não a classe

berinhard committed:
Lista eventuais warnings da planilha

berinhard committed:
Remove campos somente de leitura no momento de importação da planilha

berinhard committed:
Corrige strip de urls para limpar linha a linha

berinhard committed:
Merge branch 'feature/import-notification' into feature/spreadsheet-view

berinhard committed:
Centraliza funções para recuperar dados doGoogle Spreadsheet em um lugar

berinhard committed:
Remove função e confia nos métodos semânticos do queryset

berinhard committed:
Confia na consulta do banco de dados para recuperar apenas uma fonte de dados por estado/dia

berinhard committed:
Normaliza para usar nomes como listado no IBGE e levanta exceção no caso de typos nas planilhas do google

berinhard committed:
Cria view para servir de proxy na formatação da planilha de atualização de dados

berinhard committed:
Desativa o cache geral da view de proxy

berinhard committed:
Merge branch 'feature/spreadsheet-view' into feature/import-notification

turicas commented on covid19/spreadsheet.py:
```diff
@@ -11,11 +11,8 @@
-
     cases, reports = {}, []
-    for spreadsheet in spreadsheets:
+    for spreadsheet in StateSpreadsheet.objects.deployable_for_state(state):
```
Fazer isso adiciona um problema: não estamos checando se as URLs dos boletins compartilhadas por ambas pessoas são iguais (o ideal é não checar pois pode ser que uma pessoa tenha conseguido informações mais detalhadas).
O que sugiro fazermos é o seguinte: `final_report_urls = list(set(url for url in spreadsheet.boletim_urls for spreadsheet in spreadsheets))`, onde `spreadsheets` é todas `deployed` para aquele `state` naquela `date`. Da mesma forma, poderíamos concatenar as `boletim_notes` (`"\n".join(spreadsheet.boletim_notes for spreadsheet in spreadsheets)`).

berinhard replied:
Show! Tive que resolver isso em 2 commits, um pra acertar o banco e outro pra acertar a importação:

- 484b272219c3eb35bfe826729cd81552796646c0
- 157c22b9b1efb81ff9ab9bd7fd35698b1c772ca2

Daí fiz um teste local reproduzindo o seguinte cenário:

1. Revisão válida para o dia 21/04/2020;
2. A URL `http://google.com` estava presente nas duas importações, mas cada uma também possui uma URL exclusiva;
3. As notas de uma era `Notas do data2'` (sim, ficou um typo com esse `'`) e da outra `Notas oificiais`;

Após fazer o merge, o resultado ficou o seguinte no XLSX:

![Screenshot from 2020-04-23 12-38-22](https://user-images.githubusercontent.com/238223/80119347-f52d5e80-855f-11ea-8ec3-cb549bbf1f0c.png)

turicas replied:
Mas as notas não deveriam ser iguais (merge das duas) para todas as entradas finais de boletim?

turicas replied:
Ah, esquece. Agora entendi, vc só colocou para as que possuem URL igual. Muito bom (melhor ainda do que eu propus)!

turicas commented on covid19/spreadsheet.py:
```diff
@@ -63,17 +67,26 @@ def merge_state_data(state):
         row = row.copy()
         city = row["municipio"]
+        if city not in [TOTAL_LABEL, UNDEFINED_LABEL]:
+            city_info = get_city_info(city, state)
```
Boa!

turicas commented on covid19/spreadsheet.py:
```diff
@@ -98,4 +111,4 @@ def create_merged_state_spreadsheet(state):
     data = create_merged_state_spreadsheet("AC")
     with open("acre.xlsx", mode="wb") as fobj:
-       fobj.write(data.read())
+        fobj.write(data.read())
```
Ooops hehe

turicas commented on covid19/spreadsheet.py:
```diff
@@ -2,11 +2,15 @@
+
+
+TOTAL_LABEL = "TOTAL NO ESTADO"
+UNDEFINED_LABEL = "Importados/Indefinidos"
```
Daria para usar essas mesmas variáveis no módulo que faz a validação também? Só pra termos essas strings num lugar só.

berinhard replied:
Po que bom que vc pegou isso! Eu já tinha criado essas mesmas constantes no processo de importação, mas havia esquecido. Resolvido em 5945b30

turicas commented on covid19/views.py:
```diff
@@ -81,3 +83,18 @@ def dashboard(request):
     )
+
+
+def import_spreadsheet_provy(request, state):
```
pro**x**y?

berinhard replied:
Resolvido em b7a33e1

turicas commented on covid19/views.py:
```diff
@@ -85,6 +86,7 @@ def dashboard(request):
     )
 
 
+@disable_non_logged_user_cache
```
Ficou lindo isso! Será que vale um PR pro Django? Hehe

berinhard replied:
Honestamente, acho que vale uma libzinha pequena específica. A lib só adiciona esse middleware e pronto =)

turicas reviewed (changes requested):
Além dos comentários que fiz, vi que os testes estão falhando no CI - acho que por conta dos arquivos estáticos. Acha que vale configurarmos para não usar o S3 em ambiente de testes?

berinhard committed:
Corrige typo

berinhard committed:
Atualiza o CI para rodar o collectstatic agora por conta da renderização do 404 nos testes

berinhard committed:
Reutiliza constantes já definidas no processo de validação de planilhas

berinhard committed:
Refatora criando constantes para nomes de planilhas

berinhard committed:
Considera as urls do par de planilhas validadas e não de apenas uma delas agrupando as notas

berinhard committed:
Atualiza também o par que foi usado na revisão como DEPLOYED
